File: iOS/ACMEShop/Category.swift
//
//  Category.swift
//  ACMEShop
//
//  Copyright © 2016-present Augment. All rights reserved.
//

import Foundation

/**
 * This model is part of the Fake data model that we use to populate the app
 * Each Category has an unique id, a title and a picture, nothing special
 */
struct Category {
    
    let id: Int
    let title: String
    let picture: String
    
    /**
     * This will give the full Category list,
     * in a real app it could be an asynchronous query over the network.
     * We are keeping thing simple in this example.
     *
     * @return Array of Category
     */
    static func GetCategories() -> [Category] {
        return data.categories.map { $1 }
    }
    
    /**
     * This will give the Product list for a given Category
     * in a real app it could be an asynchronous query over the network.
     * We are keeping thing simple in this example.
     *
     * @param category the category you want the product of
     * @return Array of Product for that Category
     */
    static func GetProducts(forCategory category: Category) -> [Product] {
        var products = [Product]()
        for p in data.products_categories[category.id] ?? [] {
            if let product = data.products[p] {
                products.append(product)
            }
        }
        
        return products
    }
    
}

File: Android/app/src/main/java/com/augment/acmeshop/models/Category.java
package com.augment.acmeshop.models;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Copyright © 2016-present Augment. All rights reserved.
 */

/**
 * This model is part of the Fake data model that we use to populate the app
 * Each Category has an unique id, a title and a picture, nothing special
 */
public class Category implements Parcelable {

    private int id;
    private String title;
    private String picture;

    Category(int id, String title, String picture) {
        this.id = id;
        this.title = title;
        this.picture = picture;
    }

    /**
     * This will give the full Category list,
     * in a real app it could be an asynchronous query over the network.
     * We are keeping thing simple in this example.
     *
     * @return List of Category
     */
    public static List<Category> getCategories() {
        ArrayList<Category> categories = new ArrayList<>();
        for (Map.Entry<Integer, Category> entry : Data.categories.entrySet()) {
            categories.add(entry.getValue());
        }
        return categories;
    }

    /**
     * This will give the Product list for a given Category
     * in a real app it could be an asynchronous query over the network.
     * We are keeping thing simple in this example.
     *
     * @param category the category you want the product of
     * @return List of Product for that Category
     */
    public static List<Product> getProducts(Category category) {
        ArrayList<Product> products = new ArrayList<>();
        for (Integer p : Data.products_categories.get(category.getId())) {
            products.add(Data.products.get(p));
        }
        return products;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    @SuppressWarnings("unused")
    public void setTitle(String title) {
        this.title = title;
    }

    public String getPicture() {
        return picture;
    }

    @SuppressWarnings("unused")
    public void setPicture(String picture) {
        this.picture = picture;
    }

    // Parcelable implementation

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeStringArray(new String[]{""+id, title, picture});
    }

    public static final Creator<Category> CREATOR = new Creator<Category>() {
        @Override
        public Category createFromParcel(Parcel source) {
            String[] data = new String[3];
            source.readStringArray(data);
            return new Category(Integer.parseInt(data[0]), data[1], data[2]);
        }

        @Override
        public Category[] newArray(int size) {
            return new Category[size];
        }
    };
}
